### PR TITLE
sync APIDataSet  from kedro's `develop`

### LINKED ISSUE #183: Sync kedro.extras.datasets from develop to kedro-datasets
## Description
Short description of the problem here.
Sync kedro.extras.datasets to kedro_datasets. We found that for APIDataSet we have merged some breaking changes into `develop`, which should be sync to `kedro-datasets`.

We should also check if there are any breaking changes should go in kedro-datasets.
## Context
How has this bug affected you? What were you trying to accomplish?
To make sure we don't lose changes that has been made in the main `kedro` repo.

## Your Environment
Include as many relevant details about the environment in which you experienced the bug:

* Kedro version used (`pip show kedro` or `kedro -V`):
* Kedro plugin and kedro plugin version used (`pip show kedro-airflow`):
* Python version used (`python -V`):
* Operating system and version:

### PR DESCRIPTION
**Pull request by noklam**
## Description

Fix #183 

As per early discussion, I found `APIDataSet` and ~~`ParquetDataSet`~~ have more recent changes that we forgot to sync when we moved to `kedro-datasets`.

The sync process is quite manual, I `cp -r ` the entire directory and review one by one, tests have been updated too.
## Checklist

- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the relevant `RELEASE.md` file
- [ ] Added tests to cover my changes

### CODE REVIEW

**Review by astrojuanlu — COMMENTED** [review]
Not able to give a qualified approval 😅

**Review by merelcht — APPROVED** [review]
👍 I'd also update the release notes with these changes. They can be the same messages as originally added for the `develop` branch.

File: kedro-datasets/kedro_datasets/api/api_dataset.py [hunk 1]
```diff
@@ -34,88 +39,89 @@ class APIDataSet(AbstractDataSet[None, requests.Response]):
     data_catalog.html#use-the-data-catalog-with-the-code-api>`_:
     ::
 
-        >>> from kedro_datasets.api import APIDataSet
+        >>> from kedro.extras.datasets.api import APIDataSet
```
Comment (AhdraMeraliQB) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:42]: Saw this has already been merged but I think this got copied over by mistake, I'll change it back in #189

File: kedro-datasets/kedro_datasets/api/api_dataset.py [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
 """
-from typing import Any, Dict, Iterable, List, NoReturn, Union
+from typing import Any, Dict, List, NoReturn, Tuple, Union
 
 import requests
 from kedro.io.core import AbstractDataSet, DataSetError
+from requests import Session, sessions
 from requests.auth import AuthBase
 
+# NOTE: kedro.extras.datasets will be removed in Kedro 0.19.0.
```
Comment (AhdraMeraliQB) [inline kedro-datasets/kedro_datasets/api/api_dataset.py:11]: Same goes for this warning

**Pull request merged**